**1.Data Collection**

**1.1. Imports**

In [97]:
import pandas as pd
import numpy as np


**1.2.Load Data**

In [98]:
asthma_adults= pd.read_csv("https://raw.githubusercontent.com/Mary080102/Capston-Two/main/Raw%20data/NYC%20EH%20Data%20Portal%20-%20Asthma%20emergency%20department%20visits%20(adults)%20(full%20table).csv")
asthma_age_4= pd.read_csv("https://raw.githubusercontent.com/Mary080102/Capston-Two/main/Raw%20data/NYC%20EH%20Data%20Portal%20-%20Asthma%20emergency%20department%20visits%20(age%204%20and%20under)%20(full%20table).csv")
asthma_age5_17=pd.read_csv("https://raw.githubusercontent.com/Mary080102/Capston-Two/main/Raw%20data/NYC%20EH%20Data%20Portal%20-%20Asthma%20emergency%20department%20visits%20(age%205%20to%2017)%20(full%20table).csv")
pm2_5=pd.read_csv("https://raw.githubusercontent.com/Mary080102/Capston-Two/main/Raw%20data/NYC%20EH%20Data%20Portal%20-%20Fine%20particles%20(PM%202.5)%20(full%20table).csv")
no2=pd.read_csv("https://raw.githubusercontent.com/Mary080102/Capston-Two/main/Raw%20data/NYC%20EH%20Data%20Portal%20-%20Nitrogen%20dioxide%20(NO2)%20(full%20table).csv")

**2.Data Definition**

In [99]:
#getting a quick overview of the DataFrame
asthma_adults.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 792 entries, 0 to 791
Data columns (total 8 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   TimePeriod                        792 non-null    int64 
 1   GeoType                           792 non-null    object
 2   GeoID                             792 non-null    int64 
 3   GeoRank                           792 non-null    int64 
 4   Geography                         792 non-null    object
 5   Age-adjusted rate per 10,000      791 non-null    object
 6   Estimated annual rate per 10,000  792 non-null    object
 7   Number                            792 non-null    object
dtypes: int64(3), object(5)
memory usage: 49.6+ KB


In [100]:
asthma_age_4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 839 entries, 0 to 838
Data columns (total 7 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   TimePeriod                        839 non-null    int64 
 1   GeoType                           839 non-null    object
 2   GeoID                             839 non-null    int64 
 3   GeoRank                           839 non-null    int64 
 4   Geography                         839 non-null    object
 5   Estimated annual rate per 10,000  839 non-null    object
 6   Number                            839 non-null    object
dtypes: int64(3), object(4)
memory usage: 46.0+ KB


In [101]:
asthma_age5_17.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 839 entries, 0 to 838
Data columns (total 7 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   TimePeriod                        839 non-null    int64 
 1   GeoType                           839 non-null    object
 2   GeoID                             839 non-null    int64 
 3   GeoRank                           839 non-null    int64 
 4   Geography                         839 non-null    object
 5   Estimated annual rate per 10,000  839 non-null    object
 6   Number                            839 non-null    object
dtypes: int64(3), object(4)
memory usage: 46.0+ KB


In [102]:
pm2_5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4554 entries, 0 to 4553
Data columns (total 8 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   TimePeriod              4554 non-null   object 
 1   GeoType                 4554 non-null   object 
 2   GeoID                   4554 non-null   int64  
 3   GeoRank                 4554 non-null   int64  
 4   Geography               4554 non-null   object 
 5   10th percentile mcg/m3  4554 non-null   float64
 6   90th percentile mcg/m3  4554 non-null   float64
 7   Mean mcg/m3             4554 non-null   float64
dtypes: float64(3), int64(2), object(3)
memory usage: 284.8+ KB


In [103]:
no2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4554 entries, 0 to 4553
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   TimePeriod           4554 non-null   object 
 1   GeoType              4554 non-null   object 
 2   GeoID                4554 non-null   int64  
 3   GeoRank              4554 non-null   int64  
 4   Geography            4554 non-null   object 
 5   10th percentile ppb  4554 non-null   float64
 6   90th percentile ppb  4554 non-null   float64
 7   Mean ppb             4554 non-null   float64
dtypes: float64(3), int64(2), object(3)
memory usage: 284.8+ KB


Based on the results of .info() method, data type of some of columns must be changede. For example TimePeriod in pm2_5 dataset and no2 dataset are object that should change to int. data type of Estimated annual rate per 10,000, and Age-adjusted rate per 10,000 are object that must be foat. Data type of Number is object that must be int. After run .astype() for change data type I got error the dataset contain charector '†'. So first of all I have to find how many of this character are there? and handle it.

**2.1.Change data type of asthma_adults dataset**

In [104]:
asthma_adults["Age-adjusted rate per 10,000"].str.contains('†').sum()

1

there is exactly one entry in the "Age-adjusted rate per 10,000" column of the asthma_adults DataFrame that contains the symbol '†'.

In [105]:
# Create a copy of the DataFrame
asthma_adults_1 = asthma_adults.copy()


In [106]:
# replace '†' with empty string, then convert to numeric, setting errors='coerce'
asthma_adults_1["Age-adjusted rate per 10,000"] = pd.to_numeric(asthma_adults_1["Age-adjusted rate per 10,000"].str.replace('†', ''), errors='coerce')


In [107]:
# Calculate the mean of the column, excluding NaN values
mean_value= round(asthma_adults_1["Age-adjusted rate per 10,000"].mean(),2)

In [108]:
print(mean_value)

96.83


In [109]:
# Fill NaN values with the mean value
asthma_adults_1["Age-adjusted rate per 10,000"].fillna(mean_value, inplace=True)

In [110]:
#convert column "Estimated annual rate per 10,000" to numeric, setting errors='coerce'
asthma_adults_1["Estimated annual rate per 10,000"] = pd.to_numeric(asthma_adults_1["Estimated annual rate per 10,000"], errors='coerce')


In [111]:
asthma_adults_1["Estimated annual rate per 10,000"].isna().sum()

1

There is one entry with NaN value. I want to replace it with mean value.

In [112]:
# Calculate the mean of the column, excluding NaN values
mean_value_es=round(asthma_adults_1["Estimated annual rate per 10,000"].mean(),2)

In [113]:
print(mean_value_es)

95.58


In [114]:
# Fill NaN values with the mean value
asthma_adults_1["Estimated annual rate per 10,000"].fillna(mean_value_es, inplace=True)

In [115]:
# Using .value_counts()to see if there are any unusual values in column"Number". 
asthma_adults_1["Number"].value_counts()

520       4
494       3
982       3
1,348     3
740       3
         ..
21,089    1
20,608    1
12,872    1
9,598     1
62,425    1
Name: Number, Length: 691, dtype: int64

The result showed that some of entries has (,). I ahve to remove the comma and then change the type od data.

In [116]:
#Remove the ','  from the 'Number' in asthma_adults_1 dataset.
asthma_adults_1["Number"]= asthma_adults_1["Number"].str.replace(",", "")

In [117]:
#convert column "Number" to numeric, setting errors='coerce'
asthma_adults_1["Number"]= pd.to_numeric(asthma_adults_1["Number"], errors="coerce")

In [118]:
asthma_adults_1["Number"].isna().sum()

1

Result shows there is one entry with NaN value and it will be replace with mean value.

In [119]:
#Calculate the mean of the column "Number", excluding NaN values
mean_value_num= round(asthma_adults_1["Number"].mean(),2)

In [120]:
print(mean_value_num)

3608.03


In [121]:
# Fill NaN values with the mean value
asthma_adults_1["Number"].fillna(mean_value_num, inplace=True)

**2.2.Change data type of asthma_age_4 dataset**

Based on the results of .info(), columns of " Estimated annual rate per 10,000" and "Number" must be change to numeric data type.

In [122]:
# Create a copy of the DataFrame
asthma_age_4_1=asthma_age_4.copy()

In [123]:
# Using .value_counts()to see if there are any unusual values in column"Estimated annual rate per 10,000". 
asthma_age_4_1["Estimated annual rate per 10,000"].value_counts()

23.6     3
99.1     3
72.3     2
188.8    2
155.5    2
        ..
316.4    1
183.2    1
198.7    1
284.7    1
377.2    1
Name: Estimated annual rate per 10,000, Length: 784, dtype: int64

In [124]:
#convert column to numeric, setting errors='coerce'
asthma_age_4_1["Estimated annual rate per 10,000"]= pd.to_numeric(asthma_age_4_1["Estimated annual rate per 10,000"], errors="coerce")

In [125]:
asthma_age_4_1["Estimated annual rate per 10,000"].isna().sum()

16

Result shows that there 16 entry with NaN values in column "Estimated annual rate per 10,000". These have to be replace with mean value.

In [126]:
#Calculate the mean of the column "Estimated annual rate per 10,000", excluding NaN values
mean_age_4= round(asthma_age_4_1["Estimated annual rate per 10,000"].mean(),2)

In [127]:
print(mean_age_4)

270.35


In [128]:
# Fill NaN values with the mean value
asthma_age_4_1["Estimated annual rate per 10,000"].fillna(mean_age_4, inplace=True)

In [129]:
asthma_age_4_1["Estimated annual rate per 10,000"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 839 entries, 0 to 838
Series name: Estimated annual rate per 10,000
Non-Null Count  Dtype  
--------------  -----  
839 non-null    float64
dtypes: float64(1)
memory usage: 6.7 KB


Result shows Dtype of "Estimated annual rate per 10,000" column change to float.

In [130]:
# Using .value_counts()to see if there are any unusual values in column"Number".
asthma_age_4_1["Number"].value_counts()

122       8
118       7
50        7
23        6
43        6
         ..
1,024     1
1,454     1
1,070     1
1,561     1
19,927    1
Name: Number, Length: 543, dtype: int64

It looks like the column has number with comma, so I have to remove from data entry and then change the data type.

In [131]:
#Remove the comma from the 'Number' column
asthma_age_4_1["Number"]= asthma_age_4_1["Number"].str.replace(",", "")

In [136]:
#convert column to numeric, setting errors='coerce'
asthma_age_4_1["Number"]= pd.to_numeric(asthma_age_4_1["Number"], errors="coerce")

In [137]:
asthma_age_4_1["Number"].isna().sum()

7

There are 7 entry with NaN values that have to be replace with mean value.

In [138]:
#Calculate the mean of the column "Number", excluding NaN values
mean_age4_num= round(asthma_age_4_1["Number"].mean(),2)

In [139]:
print(mean_age4_num)

972.28


In [140]:
# Fill NaN values with the mean value
asthma_age_4_1["Number"].fillna(mean_age4_num, inplace= True)

In [141]:
asthma_age_4_1["Number"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 839 entries, 0 to 838
Series name: Number
Non-Null Count  Dtype  
--------------  -----  
839 non-null    float64
dtypes: float64(1)
memory usage: 6.7 KB


Result shows Dtype of "Number" has changed to float.

**2.3.Change data type of asthma_age5_17 dataset**

 Columns "Estimated annual rate per 10,000" and "Number" must change to numeric data type.

In [142]:
# Create a copy of the DataFrame
asthma_age5_17_1= asthma_age5_17.copy()

In [143]:
# Using .value_counts()to see if there are any unusual values in column"Estimated annual rate per 10,000".
asthma_age5_17_1["Estimated annual rate per 10,000"].value_counts()

139.9    3
47.8     3
34.6     3
98.5     3
160.0    3
        ..
78.3     1
22.4     1
120.2    1
186.4    1
453.7    1
Name: Estimated annual rate per 10,000, Length: 737, dtype: int64

In [144]:
#convert column to numeric, setting errors='coerce'
asthma_age5_17_1["Estimated annual rate per 10,000"]= pd.to_numeric(asthma_age5_17_1["Estimated annual rate per 10,000"], errors="coerce")


In [145]:
asthma_age5_17_1["Estimated annual rate per 10,000"].isna().sum()

6

There are 6 entries with NaN values that have to be replace with the mean.

In [146]:
#Calculate the mean of the column "Number", excluding NaN values
mean_5_17= round(asthma_age5_17_1["Estimated annual rate per 10,000"].mean(),2)

In [147]:
print(mean_5_17)

170.61


In [148]:
# Fill NaN values with the mean value
asthma_age5_17_1["Estimated annual rate per 10,000"].fillna(mean_5_17, inplace=True)

In [149]:
# Using .value_counts()to see if there are any unusual values in column"Estimated annual rate per 10,000".
asthma_age5_17_1["Number"].value_counts().head(20)

109      6
113      5
5*       4
29       4
135      4
116      4
139      4
253      4
202      4
81       4
94       4
85       4
327      3
351      3
37       3
244      3
1,209    3
321      3
348      3
291      3
Name: Number, dtype: int64

Results shows that some number are with * and some of them have comma. (*) in dataset means that **Estimate is based on small numbers so should be interpreted with caution**. So I want to add new column to asthma_age5_17_1 that will indicate whether the original value had an asterisk. This way, I can maintain the information about the reliability of the estimate.

In [150]:
#Create a new column to indicate estimates based on small numbers
asthma_age5_17_1["small_number_estimate"]= asthma_age5_17_1["Number"].str.contains("\*")

In [151]:
asthma_age5_17_1.head()

TimePeriod GeoType  GeoID  GeoRank                            Geography  \
0        2020      CD    101        6             Financial District (CD1)   
1        2020      CD    102        6     Greenwich Village and Soho (CD2)   
2        2020      CD    103        6  Lower East Side and Chinatown (CD3)   
3        2020      CD    104        6            Clinton and Chelsea (CD4)   
4        2020      CD    105        6                        Midtown (CD5)   

   Estimated annual rate per 10,000 Number  small_number_estimate  
0                            170.61     6*                   True  
1                            170.61     5*                   True  
2                             81.90    136                  False  
3                             42.10     34                  False  
4                            170.61     5*                   True

New column "small_number_estimate" added to the dataset and shows which number's estimate is based on small numbers so should be interpreted with caution.

In [152]:
# Remove the asterisks and commas
asthma_age5_17_1["Number"]=asthma_age5_17_1["Number"].str.replace("*", "")


/var/folders/4m/z4mg17k15k1955vh989t499c0000gn/T/ipykernel_65200/2827648353.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  asthma_age5_17_1["Number"]=asthma_age5_17_1["Number"].str.replace("*", "")


In [153]:
asthma_age5_17_1["Number"]=asthma_age5_17_1["Number"].str.replace("*", "",regex=False).str.replace(",","")

In [154]:
asthma_age5_17_1["Number"].value_counts()

109      6
113      5
5        4
29       4
135      4
        ..
228      1
1052     1
444      1
16       1
25834    1
Name: Number, Length: 614, dtype: int64

In [155]:
#convert column to numeric, setting errors='coerce'
asthma_age5_17_1["Number"]= pd.to_numeric(asthma_age5_17_1["Number"], errors="coerce")

In [156]:
asthma_age5_17_1["Number"].isna().sum()

0

**2.4.Change data type of pm2_5 dataset**

The "TimePeriod" column has object data type that has to change to numeric.

In [157]:
# Creat a copy of dataFrame
pm2_5_1= pm2_5.copy()

In [158]:
# Using .value_counts()to see if there are any unusual values in column"TimePeriod".
pm2_5_1["TimePeriod"].value_counts()

Annual Average 2022    141
Summer 2018            141
Winter 2014-15         141
Summer 2015            141
Summer 2022            141
Winter 2015-16         141
Summer 2016            141
Annual Average 2016    141
Winter 2016-17         141
Summer 2017            141
Annual Average 2017    141
Winter 2017-18         141
Annual Average 2015    141
Annual Average 2018    141
Annual Average 2020    141
Winter 2018-19         141
Winter 2021-22         141
Summer 2021            141
Winter 2020-21         141
Annual Average 2021    141
Summer 2020            141
Winter 2019-20         141
Annual Average 2019    141
Summer 2019            141
Annual Average 2011     65
Summer 2009             65
Annual Average 2009     65
Winter 2009-10          65
Summer 2010             65
Annual Average 2010     65
Winter 2010-11          65
Summer 2011             65
Annual Average 2013     65
Winter 2011-12          65
Summer 2012             65
Annual Average 2012     65
Winter 2012-13          65
S

Results of .value_counts() shows different type of entry including: Annual Average, Summer, and Winter. Because values of the other dataframes sucha as asthma_adults, asthma_age_4, and asthma_5_17 are based on annual not seasonal, therefore I decide to keep just Annual Average values.

In [159]:
# filter rows to keep Annual Average rows
pm2_5_sub= pm2_5_1[pm2_5_1["TimePeriod"].str.startswith("Annual Average")]

In [160]:
pm2_5_sub.head()

TimePeriod GeoType  GeoID  GeoRank  \
0  Annual Average 2022      CD    101        6   
1  Annual Average 2022      CD    102        6   
2  Annual Average 2022      CD    103        6   
3  Annual Average 2022      CD    104        6   
4  Annual Average 2022      CD    105        6   

                             Geography  10th percentile mcg/m3  \
0             Financial District (CD1)                     6.7   
1     Greenwich Village and Soho (CD2)                     7.6   
2  Lower East Side and Chinatown (CD3)                     6.1   
3            Clinton and Chelsea (CD4)                     6.7   
4                        Midtown (CD5)                     8.2   

   90th percentile mcg/m3  Mean mcg/m3  
0                     7.8          7.2  
1                     9.3          8.4  
2                     8.6          7.2  
3                     9.0          7.8  
4                     9.7          9.1

All data for "TimePeriod" column just include Annual Average.

In [161]:
#Remove the 'Annual' prefix from the 'TimePeriod' in pm2_5_sub
pm2_5_sub= pm2_5_sub.copy()
pm2_5_sub["TimePeriod"]= pm2_5_sub["TimePeriod"].str.replace("Annual Average", "")

In [162]:
pm2_5_sub.head()

TimePeriod GeoType  GeoID  GeoRank                            Geography  \
0       2022      CD    101        6             Financial District (CD1)   
1       2022      CD    102        6     Greenwich Village and Soho (CD2)   
2       2022      CD    103        6  Lower East Side and Chinatown (CD3)   
3       2022      CD    104        6            Clinton and Chelsea (CD4)   
4       2022      CD    105        6                        Midtown (CD5)   

   10th percentile mcg/m3  90th percentile mcg/m3  Mean mcg/m3  
0                     6.7                     7.8          7.2  
1                     7.6                     9.3          8.4  
2                     6.1                     8.6          7.2  
3                     6.7                     9.0          7.8  
4                     8.2                     9.7          9.1

In [163]:
#Creat a copy of dataframe
no2_1= no2.copy()

In [164]:
# filter rows to keep Annual Average rows
no2_sub= no2_1[no2_1["TimePeriod"].str.startswith("Annual Average")]
no2_sub.head()

TimePeriod GeoType  GeoID  GeoRank  \
0  Annual Average 2022      CD    101        6   
1  Annual Average 2022      CD    102        6   
2  Annual Average 2022      CD    103        6   
3  Annual Average 2022      CD    104        6   
4  Annual Average 2022      CD    105        6   

                             Geography  10th percentile ppb  \
0             Financial District (CD1)                 19.3   
1     Greenwich Village and Soho (CD2)                 19.2   
2  Lower East Side and Chinatown (CD3)                 15.8   
3            Clinton and Chelsea (CD4)                 18.6   
4                        Midtown (CD5)                 22.4   

   90th percentile ppb  Mean ppb  
0                 22.4      21.0  
1                 21.1      20.2  
2                 19.9      17.8  
3                 24.4      21.4  
4                 27.8      25.5

In [165]:
#Remove the 'Annual' prefix from the 'TimePeriod' in pm2_5_sub
no2_sub=no2_sub.copy()
no2_sub["TimePeriod"]= no2_sub["TimePeriod"].str.replace("Annual Average", "")
no2_sub.head()

TimePeriod GeoType  GeoID  GeoRank                            Geography  \
0       2022      CD    101        6             Financial District (CD1)   
1       2022      CD    102        6     Greenwich Village and Soho (CD2)   
2       2022      CD    103        6  Lower East Side and Chinatown (CD3)   
3       2022      CD    104        6            Clinton and Chelsea (CD4)   
4       2022      CD    105        6                        Midtown (CD5)   

   10th percentile ppb  90th percentile ppb  Mean ppb  
0                 19.3                 22.4      21.0  
1                 19.2                 21.1      20.2  
2                 15.8                 19.9      17.8  
3                 18.6                 24.4      21.4  
4                 22.4                 27.8      25.5

In [166]:
# Change the data type into int
pm2_5_sub["TimePeriod"]=pm2_5_sub["TimePeriod"].astype(int)

no2_sub["TimePeriod"]=no2_sub["TimePeriod"].astype(int)

In [167]:

no2_sub["TimePeriod"].info()

<class 'pandas.core.series.Series'>
Int64Index: 1518 entries, 0 to 4423
Series name: TimePeriod
Non-Null Count  Dtype
--------------  -----
1518 non-null   int64
dtypes: int64(1)
memory usage: 23.7 KB


Data type on the "TimePeriod" column changed to int.

**3.Data Cleaning**

**3.1. Dropping Unneccessary Rows:**

I want to choose three neighborhoods in New York City to assess air quality that can provide an overview of the whole area requires considering diversity in geography, urban density, and potential sources of pollution. Based on the Geography column in csv files that includes all neighborhoods,I want to choose New yourk, Manhatan,Staten Island.

In [168]:
# List of names I want to keep
rows_to_keep=["New York City","Manhattan","Staten Island"]

In [169]:
# Creat a new Dataframe with only rows for our study with the name,"New York City","Manhattan","Staten Island"
asthma_adults_df=asthma_adults_1.loc[asthma_adults_1["Geography"].isin(rows_to_keep)]

asthma_age_4_df= asthma_age_4_1.loc[asthma_age_4_1["Geography"].isin(rows_to_keep)]

asthma_age5_17_df=asthma_age5_17_1.loc[asthma_age5_17_1["Geography"].isin(rows_to_keep)]

pm2_5_df= pm2_5_sub.loc[pm2_5_sub["Geography"].isin(rows_to_keep)]

no2_df= no2_sub.loc[no2_sub["Geography"].isin(rows_to_keep)]

**3.2.Dropping unneccessary Columns**

In [170]:
# finding common column to merge
common_columns= set(asthma_adults_df.columns)& set(asthma_age_4_df.columns)&set(asthma_age5_17_df.columns)&set(pm2_5_df.columns)&set(no2_df.columns)
print(common_columns)

{'GeoType', 'GeoID', 'Geography', 'GeoRank', 'TimePeriod'}


These are common columns in all five dataFrames. I can choose one of them as a key column to merge all datasets. I want to choose "GeoID" as a key column and drop "GeoType" and "GeoRank" columns. Because these column doesn't have any valuable data.

In [171]:
# List the columns that want to drop
columns_to_drop=["GeoType","GeoRank"]


In [172]:
# Create a new DataFrame with only the necessary columns

asthma_adults_df1= asthma_adults_df.drop(columns_to_drop, axis=1)

asthma_age_4_df1=asthma_age_4.drop(columns_to_drop,axis=1)

asthma_age5_17_df1=asthma_age5_17_df.drop(columns_to_drop, axis=1)

pm2_5_df1= pm2_5_df.drop(columns_to_drop,axis=1)

no2_df1= no2_df.drop(columns_to_drop,axis=1)

In [173]:
print(asthma_adults_df1.columns)


Index(['TimePeriod', 'GeoID', 'Geography', 'Age-adjusted rate per 10,000',
       'Estimated annual rate per 10,000', 'Number'],
      dtype='object')


Columns "GeoType" and "GeoRank" have removed from dataframe.

In [174]:
#rename the column 'TimePeriod' to 'Year'
asthma_adults_df1= asthma_adults_df1.rename(columns={"TimePeriod":"Year"})

asthma_age_4_df1=asthma_age_4_df1.rename(columns={"TimePeriod":"Year"})

asthma_age5_17_df1=asthma_age5_17_df1.rename(columns={"TimePeriod":"Year"})

pm2_5_df1= pm2_5_df1.rename(columns={"TimePeriod":"Year"})

no2_df1= no2_df1.rename(columns={"TimePeriod":"Year"})

In [175]:
no2_df1.head()

Year  GeoID      Geography  10th percentile ppb  90th percentile ppb  \
137  2022      3      Manhattan                 16.0                 23.3   
139  2022      5  Staten Island                  8.4                 13.9   
140  2022      1  New York City                 10.8                 18.1   
560  2021      3      Manhattan                 17.0                 23.7   
562  2021      5  Staten Island                 10.1                 15.8   

     Mean ppb  
137      19.1  
139      11.2  
140      14.7  
560      20.0  
562      12.8

**4.Joining Data**

In [176]:
# rename the common columns except 'GeoID' to merge dataframes
asthma_adults_df1= asthma_adults_df1.rename(columns={'Geography':'Geography_ad',"Year":"Year_ad"})

asthma_age_4_df1= asthma_age_4_df1.rename(columns={'Geography':'Geography_ag_4',"Year":"Year_ag_4"})

asthma_age5_17_df1= asthma_age5_17_df1.rename(columns={'Geography':'Geography_ag5_17',"Year":"Year5_17"})

no2_df1=no2_df1.rename(columns={'Geography':'Geography_no2',"Year":"Year_no"})

pm2_5_df1=pm2_5_df1.rename(columns={'Geography':'Geography_pm', "Year":"Year_pm"})

In [177]:
merge1=pd.merge(asthma_adults_df1,asthma_age_4_df1, on="GeoID", how="inner")


In [178]:
merge2= pd.merge(merge1,asthma_age5_17_df1, on="GeoID", how="inner")

In [179]:
merge3= pd.merge(merge2,pm2_5_df1, on="GeoID", how="inner")

In [180]:
merge4=pd.merge(merge3,no2_df1, on="GeoID", how="inner")

In [181]:
merge4.head()

Year_ad  GeoID Geography_ad  Age-adjusted rate per 10,000  \
0     2020      3    Manhattan                          44.7   
1     2020      3    Manhattan                          44.7   
2     2020      3    Manhattan                          44.7   
3     2020      3    Manhattan                          44.7   
4     2020      3    Manhattan                          44.7   

   Estimated annual rate per 10,000_x  Number_x  Year_ag_4 Geography_ag_4  \
0                                43.5    6290.0       2020      Manhattan   
1                                43.5    6290.0       2020      Manhattan   
2                                43.5    6290.0       2020      Manhattan   
3                                43.5    6290.0       2020      Manhattan   
4                                43.5    6290.0       2020      Manhattan   

  Estimated annual rate per 10,000_y Number_y  ...  Year_pm Geography_pm  \
0                               58.6      454  ...     2022    Manhattan   
1                               58.6      454  ...     2022    Manhattan   
2                               58.6      454  ...     2022    Manhattan   
3                               58.6      454  ...     2022    Manhattan   
4                               58.6      454  ...     2022    Manhattan   

   10th percentile mcg/m3  90th percentile mcg/m3  Mean mcg/m3  Year_no  \
0                     6.1                     8.8          7.0     2022   
1                     6.1                     8.8          7.0     2021   
2                     6.1                     8.8          7.0     2020   
3                     6.1                     8.8          7.0     2019   
4                     6.1                     8.8          7.0     2018   

  Geography_no2  10th percentile ppb  90th percentile ppb  Mean ppb  
0     Manhattan                 16.0                 23.3      19.1  
1     Manhattan                 17.0                 23.7      20.0  
2     Manhattan                 17.7                 24.4      20.5  
3     Manhattan                 17.7                 26.6      21.7  
4     Manhattan                 19.4                 28.8      23.4  

[5 rows x 25 columns]

Results show some duplicated columns that have to be remove.

In [182]:
# Duplicated columns to drop
columns_to_remove=["Year_ag_4","Geography_ag_4","Year_pm","Geography_pm","Year_no","Geography_no2",'Geography_ag5_17',"Year5_17"]

In [183]:
merge4= merge4.drop(columns_to_remove, axis=1)

In [184]:
merge4.head()

Year_ad  GeoID Geography_ad  Age-adjusted rate per 10,000  \
0     2020      3    Manhattan                          44.7   
1     2020      3    Manhattan                          44.7   
2     2020      3    Manhattan                          44.7   
3     2020      3    Manhattan                          44.7   
4     2020      3    Manhattan                          44.7   

   Estimated annual rate per 10,000_x  Number_x  \
0                                43.5    6290.0   
1                                43.5    6290.0   
2                                43.5    6290.0   
3                                43.5    6290.0   
4                                43.5    6290.0   

  Estimated annual rate per 10,000_y Number_y  \
0                               58.6      454   
1                               58.6      454   
2                               58.6      454   
3                               58.6      454   
4                               58.6      454   

   Estimated annual rate per 10,000  Number  small_number_estimate  \
0                              65.9    1088                  False   
1                              65.9    1088                  False   
2                              65.9    1088                  False   
3                              65.9    1088                  False   
4                              65.9    1088                  False   

   10th percentile mcg/m3  90th percentile mcg/m3  Mean mcg/m3  \
0                     6.1                     8.8          7.0   
1                     6.1                     8.8          7.0   
2                     6.1                     8.8          7.0   
3                     6.1                     8.8          7.0   
4                     6.1                     8.8          7.0   

   10th percentile ppb  90th percentile ppb  Mean ppb  
0                 16.0                 23.3      19.1  
1                 17.0                 23.7      20.0  
2                 17.7                 24.4      20.5  
3                 17.7                 26.6      21.7  
4                 19.4                 28.8      23.4

In [207]:
# Rename some of column's name
merge4= merge4.rename(columns={"Year_ad":"Year", "Geography_ad":"Geography","Age-adjusted rate per 10,000":"ad_per 10,000",
                              "Estimated annual rate per 10,000_x":"annual_ad per 10,000 ","Number_x":"Number_ad",
                               "Number_y":"Number_ag4", "Estimated annual rate per 10,000_y":"annual_ag4 per 10,000","Estimated annual rate per 10,000":"annual5_17 per 10,000",
                              "Number":"Number5_17","Mean mcg/m3":"pm/Mean mcg/m3","Mean ppb":"no2/Mean ppb"})

In [208]:
merge4.head()

Year  GeoID  Geography  ad_per 10,000  annual_ad per 10,000   Number_ad  \
0  2020      3  Manhattan           44.7                   43.5     6290.0   
1  2020      3  Manhattan           44.7                   43.5     6290.0   
2  2020      3  Manhattan           44.7                   43.5     6290.0   
3  2020      3  Manhattan           44.7                   43.5     6290.0   
4  2020      3  Manhattan           44.7                   43.5     6290.0   

   annual_ag4 per 10,000  Number_ag4  annual5_17 per 10,000  Number5_17  \
0                   58.6         454                   65.9        1088   
1                   58.6         454                   65.9        1088   
2                   58.6         454                   65.9        1088   
3                   58.6         454                   65.9        1088   
4                   58.6         454                   65.9        1088   

   small_number_estimate  10th percentile mcg/m3  90th percentile mcg/m3  \
0                  False                     6.1                     8.8   
1                  False                     6.1                     8.8   
2                  False                     6.1                     8.8   
3                  False                     6.1                     8.8   
4                  False                     6.1                     8.8   

   pm/Mean mcg/m3  10th percentile ppb  90th percentile ppb  no2/Mean ppb  
0             7.0                 16.0                 23.3          19.1  
1             7.0                 17.0                 23.7          20.0  
2             7.0                 17.7                 24.4          20.5  
3             7.0                 17.7                 26.6          21.7  
4             7.0                 19.4                 28.8          23.4

In [209]:
merge4.shape

(2790256, 17)

In [210]:
# Using .ifo() to check Dtype of our data
merge4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2790256 entries, 0 to 2790255
Data columns (total 17 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Year                    int64  
 1   GeoID                   int64  
 2   Geography               object 
 3   ad_per 10,000           float64
 4   annual_ad per 10,000    float64
 5   Number_ad               float64
 6   annual_ag4 per 10,000   float64
 7   Number_ag4              int64  
 8   annual5_17 per 10,000   float64
 9   Number5_17              int64  
 10  small_number_estimate   bool   
 11  10th percentile mcg/m3  float64
 12  90th percentile mcg/m3  float64
 13  pm/Mean mcg/m3          float64
 14  10th percentile ppb     float64
 15  90th percentile ppb     float64
 16  no2/Mean ppb            float64
dtypes: bool(1), float64(11), int64(4), object(1)
memory usage: 364.6+ MB


In [211]:
# calculates the sum of NA (missing) values.
merge4.isna().sum()

Year                      0
GeoID                     0
Geography                 0
ad_per 10,000             0
annual_ad per 10,000      0
Number_ad                 0
annual_ag4 per 10,000     0
Number_ag4                0
annual5_17 per 10,000     0
Number5_17                0
small_number_estimate     0
10th percentile mcg/m3    0
90th percentile mcg/m3    0
pm/Mean mcg/m3            0
10th percentile ppb       0
90th percentile ppb       0
no2/Mean ppb              0
dtype: int64

There is no any NaN values in the dataframe.

In [212]:
# calculates the sum of duplicate values
merge4.duplicated().sum()

111104

There are 111104 duplicate data in merged dataset.

**5.Numerical Features**

In [213]:
# Filter columns with number data to claculate some statistic such as: mean,max,min,...
merge4.select_dtypes(include=["number"])

Year  GeoID  ad_per 10,000  annual_ad per 10,000   Number_ad  \
0        2020      3           44.7                   43.5     6290.0   
1        2020      3           44.7                   43.5     6290.0   
2        2020      3           44.7                   43.5     6290.0   
3        2020      3           44.7                   43.5     6290.0   
4        2020      3           44.7                   43.5     6290.0   
...       ...    ...            ...                    ...        ...   
2790251  2005      1           99.5                  100.3    62425.0   
2790252  2005      1           99.5                  100.3    62425.0   
2790253  2005      1           99.5                  100.3    62425.0   
2790254  2005      1           99.5                  100.3    62425.0   
2790255  2005      1           99.5                  100.3    62425.0   

         annual_ag4 per 10,000  Number_ag4  annual5_17 per 10,000  Number5_17  \
0                         58.6         454                   65.9        1088   
1                         58.6         454                   65.9        1088   
2                         58.6         454                   65.9        1088   
3                         58.6         454                   65.9        1088   
4                         58.6         454                   65.9        1088   
...                        ...         ...                    ...         ...   
2790251                  377.2       19927                  199.2       25834   
2790252                  377.2       19927                  199.2       25834   
2790253                  377.2       19927                  199.2       25834   
2790254                  377.2       19927                  199.2       25834   
2790255                  377.2       19927                  199.2       25834   

         10th percentile mcg/m3  90th percentile mcg/m3  pm/Mean mcg/m3  \
0                           6.1                     8.8             7.0   
1                           6.1                     8.8             7.0   
2                           6.1                     8.8             7.0   
3                           6.1                     8.8             7.0   
4                           6.1                     8.8             7.0   
...                         ...                     ...             ...   
2790251                     9.3                    11.9            10.4   
2790252                     9.3                    11.9            10.4   
2790253                     9.3                    11.9            10.4   
2790254                     9.3                    11.9            10.4   
2790255                     9.3                    11.9            10.4   

         10th percentile ppb  90th percentile ppb  no2/Mean ppb  
0                       16.0                 23.3          19.1  
1                       17.0                 23.7          20.0  
2                       17.7                 24.4          20.5  
3                       17.7                 26.6          21.7  
4                       19.4                 28.8          23.4  
...                      ...                  ...           ...  
2790251                 12.6                 23.6          18.8  
2790252                 13.3                 24.3          19.6  
2790253                 14.7                 26.7          21.6  
2790254                 14.1                 26.3          21.1  
2790255                 15.4                 28.6          22.7  

[2790256 rows x 15 columns]

Result of .select_dtypes() shows that two columns of our data("annual_ag4 per 10,000" and "Number_ag4") did not chane into numeric.

In [214]:
# #convert column to numeric
merge4["annual_ag4 per 10,000"]=pd.to_numeric(merge4["annual_ag4 per 10,000"])

In [215]:
# # Using .ifo() to check Dtype of our data
merge4["annual_ag4 per 10,000"].info()

<class 'pandas.core.series.Series'>
Int64Index: 2790256 entries, 0 to 2790255
Series name: annual_ag4 per 10,000
Non-Null Count    Dtype  
--------------    -----  
2790256 non-null  float64
dtypes: float64(1)
memory usage: 42.6 MB


In [216]:
merge4["Number_ag4"].info()

<class 'pandas.core.series.Series'>
Int64Index: 2790256 entries, 0 to 2790255
Series name: Number_ag4
Non-Null Count    Dtype
--------------    -----
2790256 non-null  int64
dtypes: int64(1)
memory usage: 42.6 MB


In [217]:
merge4.select_dtypes(include=["number"])

Year  GeoID  ad_per 10,000  annual_ad per 10,000   Number_ad  \
0        2020      3           44.7                   43.5     6290.0   
1        2020      3           44.7                   43.5     6290.0   
2        2020      3           44.7                   43.5     6290.0   
3        2020      3           44.7                   43.5     6290.0   
4        2020      3           44.7                   43.5     6290.0   
...       ...    ...            ...                    ...        ...   
2790251  2005      1           99.5                  100.3    62425.0   
2790252  2005      1           99.5                  100.3    62425.0   
2790253  2005      1           99.5                  100.3    62425.0   
2790254  2005      1           99.5                  100.3    62425.0   
2790255  2005      1           99.5                  100.3    62425.0   

         annual_ag4 per 10,000  Number_ag4  annual5_17 per 10,000  Number5_17  \
0                         58.6         454                   65.9        1088   
1                         58.6         454                   65.9        1088   
2                         58.6         454                   65.9        1088   
3                         58.6         454                   65.9        1088   
4                         58.6         454                   65.9        1088   
...                        ...         ...                    ...         ...   
2790251                  377.2       19927                  199.2       25834   
2790252                  377.2       19927                  199.2       25834   
2790253                  377.2       19927                  199.2       25834   
2790254                  377.2       19927                  199.2       25834   
2790255                  377.2       19927                  199.2       25834   

         10th percentile mcg/m3  90th percentile mcg/m3  pm/Mean mcg/m3  \
0                           6.1                     8.8             7.0   
1                           6.1                     8.8             7.0   
2                           6.1                     8.8             7.0   
3                           6.1                     8.8             7.0   
4                           6.1                     8.8             7.0   
...                         ...                     ...             ...   
2790251                     9.3                    11.9            10.4   
2790252                     9.3                    11.9            10.4   
2790253                     9.3                    11.9            10.4   
2790254                     9.3                    11.9            10.4   
2790255                     9.3                    11.9            10.4   

         10th percentile ppb  90th percentile ppb  no2/Mean ppb  
0                       16.0                 23.3          19.1  
1                       17.0                 23.7          20.0  
2                       17.7                 24.4          20.5  
3                       17.7                 26.6          21.7  
4                       19.4                 28.8          23.4  
...                      ...                  ...           ...  
2790251                 12.6                 23.6          18.8  
2790252                 13.3                 24.3          19.6  
2790253                 14.7                 26.7          21.6  
2790254                 14.1                 26.3          21.1  
2790255                 15.4                 28.6          22.7  

[2790256 rows x 15 columns]

Based on the above table "Number_ag4" and "annual_ag4 per 10,000" have changed into number Dtype and I can use .decsribe() method to generate descriptive statistics of the data in a DataFrame.

In [218]:
#Creat descriptive statistics of number of asthma for adults 
merge4["Number_ad"].describe()

count    2.790256e+06
mean     3.780872e+04
std      2.990139e+04
min      1.287000e+03
25%      3.608030e+03
50%      3.323800e+04
75%      6.622900e+04
max      7.779200e+04
Name: Number_ad, dtype: float64

In [219]:
#Creat descriptive statistics of number of asthma for children up to 4 years.
merge4["Number_ag4"].describe()

count    2.790256e+06
mean     7.313927e+03
std      7.124873e+03
min      9.400000e+01
25%      1.300000e+03
50%      4.876000e+03
75%      1.285200e+04
max      2.149200e+04
Name: Number_ag4, dtype: float64

In [220]:
#Creat descriptive statistics of number of asthma for children aged 5 to 17 years.
merge4["Number5_17"].describe()

count    2.790256e+06
mean     1.482970e+04
std      1.151949e+04
min      2.180000e+02
25%      3.046000e+03
50%      1.903400e+04
75%      2.583400e+04
max      2.837000e+04
Name: Number5_17, dtype: float64

In [221]:
#Creat descriptive statistics of PM2.5.
merge4["pm/Mean mcg/m3"].describe()

count    2.790256e+06
mean     8.213306e+00
std      1.711961e+00
min      5.200000e+00
25%      6.800000e+00
50%      8.350000e+00
75%      9.400000e+00
max      1.260000e+01
Name: pm/Mean mcg/m3, dtype: float64

In [222]:
#Creat descriptive statistics of no2.
merge4["no2/Mean ppb"].describe()

count    2.790256e+06
mean     1.882980e+01
std      4.888561e+00
min      1.120000e+01
25%      1.520000e+01
50%      1.810000e+01
75%      2.160000e+01
max      3.390000e+01
Name: no2/Mean ppb, dtype: float64

 **6.Categorical Features**

In [223]:
merge4.select_dtypes(include=['object'])


Geography
0            Manhattan
1            Manhattan
2            Manhattan
3            Manhattan
4            Manhattan
...                ...
2790251  New York City
2790252  New York City
2790253  New York City
2790254  New York City
2790255  New York City

[2790256 rows x 1 columns]